sql/atxs/atxs.go

```diff
@@ -793,3 +772,43 @@ func IterateAtxIdsWithMalfeasance(
 	)
 	return err
 }
+
+type PrevATXCollision struct {
+	NodeID1 types.NodeID
+	ATX1    types.ATXID
+
+	NodeID2 types.NodeID
+	ATX2    types.ATXID
+}
+
+func PrevATXCollisions(db sql.Executor) ([]PrevATXCollision, error) {
+	var result []PrevATXCollision
+
+	dec := func(stmt *sql.Statement) bool {
+		var nodeID1, nodeID2 types.NodeID
+		stmt.ColumnBytes(0, nodeID1[:])
+		stmt.ColumnBytes(1, nodeID2[:])
+
+		var id1, id2 types.ATXID
+		stmt.ColumnBytes(2, id1[:])
+		stmt.ColumnBytes(3, id2[:])
+
+		result = append(result, PrevATXCollision{
+			NodeID1: nodeID1,
+			ATX1:    id1,
+
+			NodeID2: nodeID2,
+			ATX2:    id2,
+		})
+		return true
+	}
+	if _, err := db.Exec(`
+		SELECT t1.pubkey, t2.pubkey, t1.id, t2.id
+		FROM atxs t1
+		INNER JOIN atxs t2 ON t1.prev_id = t2.prev_id
+		WHERE t1.id < t2.id;`, nil, dec); err != nil {
+		return nil, fmt.Errorf("error getting ATXs with same prevATX: %w", err)
+	}
```

Comment:
> why does it pick a pubkey twice, for each ATX? An ATX pointing to a previous ATX owned by a different smesher is syntactically invalid.

It doesn't return the `pubkey` twice. It returns the `pubkey` of ATX1 and the `pubkey` of ATX2. I added this as a safety check so I don't accidentally create malfeasance proofs for ATXs referencing an ATX of a different smesher as their previous. Returning the `pubkey` from both ATXs is just as cheap compared to only returning one `pubkey`.

This query is only used in `--scan-malfeasant-atxs`, an optional startup parameter that can be removed again in `v1.7.x` (for now its only so smeshers can verify that we marked the correct IDs as malicious).

> why does it compare t1.id < t2.id?

If this wouldn't be part of the query it would return all malicious ATX pairs twice; e.g. (ATX1, ATX2) and (ATX2, ATX1). The additional `WHERE` clause ensures that only one of those is returned from the query.

Reply:
I hope there aren't any such ATXs in the DB as it would be a horrible bug.

>>why does it compare t1.id < t2.id?
>
>If this wouldn't be part of the query it would return all malicious ATX pairs twice; e.g. (ATX1, ATX2) and (ATX2, ATX1). The additional WHERE clause ensures that only one of those is returned from the query.

Interesting, but not very obvious. How about leaving a comment explaining it?